File: 6. Binary Search/528. Random Pick with Weight/Solution.py
from typing import List
import random

class Solution:

    def __init__(self, w: List[int]):
        self.prefix_sums = [w[0]]
        for i in range(1, len(w)):
            self.prefix_sums.append(self.prefix_sums[-1] + w[i])

    def pickIndex(self) -> int:
        target = random.randint(1, self.prefix_sums[-1])

        left = 0
        right = len(self.prefix_sums) - 1

        while left < right:
            middle = (right + left) // 2

            if self.prefix_sums[middle] < target:
                left = middle + 1
            else:
                right = middle
        
        return left

File: 6. Binary Search/528. Random Pick with Weight/Solution.cpp
#include <vector>
#include <cstdlib>

class Solution {
private:
    std::vector<int> prefixSums;
public:
    Solution(std::vector<int>& w) {
        prefixSums.push_back(w[0]);
        for (int i = 1; i < w.size(); ++i)
        {
            prefixSums.push_back(prefixSums.back() + w[i]);
        }
    }
    
    int pickIndex() {
        int target = std::rand() % prefixSums.back() + 1;
        int left = 0;
        int right = prefixSums.size() - 1;

        while (left < right)
        {
            int middle = (right + left) / 2;

            if (prefixSums[middle] < target)
            {
                left = middle + 1;
            }
            else
            {
                right = middle;
            }
        }

        return left;
    }
};